File: src/modules/auth/controllers/auth.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Post,
  UseGuards,
} from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger';
import { CurrentUser } from '../../../common/decorators/current-user.decorator';
import type { AuthenticatedUser } from '../../../common/types/authenticated-user.type';
import { JwtAuthGuard } from '../../auth/guards/jwt-auth.guard';
import { AuthTokenResponseDto } from '../dto/auth-token-response.dto';
import { LoginDto } from '../dto/login.dto';
import { RefreshTokenDto } from '../dto/refresh-token.dto';
import { RegisterDto } from '../dto/register.dto';
import { AuthService } from '../services/auth.service';

@ApiTags('auth')
@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('register')
  @ApiOperation({
    summary: 'Register a new employee account',
    description:
      'Creates an employee account with the default employee role. Email must be unique.',
  })
  register(@Body() registerDto: RegisterDto) {
    return this.authService.register(registerDto);
  }

  @Post('login')
  @ApiOkResponse({ type: AuthTokenResponseDto })
  @ApiOperation({
    summary: 'Authenticate and receive tokens',
    description:
      'Returns an access JWT and a refresh token. Use the access token as Bearer auth on protected routes.',
  })
  login(@Body() loginDto: LoginDto) {
    return this.authService.login(loginDto);
  }

  @Post('refresh')
  @ApiOkResponse({ type: AuthTokenResponseDto })
  @ApiUnauthorizedResponse({ description: 'Invalid or expired refresh token' })
  @ApiOperation({
    summary: 'Refresh access token',
    description:
      'Rotates the refresh token and returns a new access/refresh token pair.',
  })
  refresh(@Body() dto: RefreshTokenDto) {
    return this.authService.refresh(dto.refreshToken);
  }

  @Post('logout')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiUnauthorizedResponse({ description: 'Invalid or expired refresh token' })
  @ApiOperation({
    summary: 'Revoke refresh token',
    description:
      'Invalidates the provided refresh token. Access tokens expire naturally.',
  })
  async logout(@Body() dto: RefreshTokenDto): Promise<void> {
    await this.authService.logout(dto.refreshToken);
  }

  @ApiBearerAuth()
  @UseGuards(JwtAuthGuard)
  @Get('me')
  @ApiOperation({
    summary: 'Get the current user profile',
    description:
      'Returns the authenticated user profile including department and direct manager.',
  })
  me(@CurrentUser() user: AuthenticatedUser) {
    return this.authService.getProfile(user.id);
  }
}

File: src/EquipmentApi/Controllers/AuthController.cs
using EquipmentApi.Auth;
using EquipmentApi.Dtos;
using EquipmentApi.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace EquipmentApi.Controllers;

[ApiController]
[Route("auth")]
public class AuthController : ControllerBase
{
    private readonly AuthService _authService;

    public AuthController(AuthService authService)
    {
        _authService = authService;
    }

    [HttpPost("register")]
    public Task<Entities.Employee> Register([FromBody] RegisterDto dto, CancellationToken cancellationToken) =>
        _authService.RegisterAsync(dto, cancellationToken);

    [HttpPost("login")]
    public Task<AuthTokenResponse> Login([FromBody] LoginDto dto, CancellationToken cancellationToken) =>
        _authService.LoginAsync(dto, cancellationToken);

    [HttpPost("refresh")]
    public Task<AuthTokenResponse> Refresh([FromBody] RefreshTokenDto dto, CancellationToken cancellationToken) =>
        _authService.RefreshAsync(dto.RefreshToken, cancellationToken);

    [HttpPost("logout")]
    public async Task<IActionResult> Logout([FromBody] RefreshTokenDto dto, CancellationToken cancellationToken)
    {
        await _authService.LogoutAsync(dto.RefreshToken, cancellationToken);
        return NoContent();
    }

    [Authorize]
    [HttpGet("me")]
    public Task<Entities.Employee> Me(CancellationToken cancellationToken) =>
        _authService.GetProfileAsync(User.GetAuthenticatedUser().Id, cancellationToken);
}
